### PR TITLE
Add SPSA (Simultaneous Perturbation Stochastic Approximation) optimizer to contrib

### LINKED ISSUE #357: Add SPSA optimization method
The  Simultaneous Perturbation Stochastic Approximation (SPSA) optimisation method is a faster optimisation method.

>  If the number of terms being optimized is p, then the finite-difference method takes 2p measurements of the objective function at each iteration (to form one gradient approximation), while SPSA takes only two measurements

It is also naturally suited for noisy measurements. Thus, it will be useful when simulating noisy systems.

The theory (and implementation) for SPSA is:
* Described in James C. Spall, “[An Overview of the Simultaneous Perturbation Method for Efficient Optimization](https://www.jhuapl.edu/SPSA/PDF-SPSA/Spall_An_Overview.PDF)”, 1998;

Furthermore, it is implemented:
* In the [`noisyopt` package](https://noisyopt.readthedocs.io/en/latest/), specifically see [the source code here](https://github.com/andim/noisyopt/blob/master/noisyopt/main.py#L264);
* In Qiskit, [see the SPSA optimizer documentation](https://qiskit.org/documentation/stubs/qiskit.algorithms.optimizers.SPSA.html#qiskit.algorithms.optimizers.SPSA).

More information:
https://www.jhuapl.edu/SPSA/

### PR DESCRIPTION
**Pull request by jaideeppyne**
Resolves #357 (labeled `help wanted`).

## What

Adds **SPSA** (Simultaneous Perturbation Stochastic Approximation; Spall, 1992) to `optax.contrib`. SPSA is a **gradient-free** optimizer: it estimates the entire gradient from only **two** objective evaluations per step, independent of the number of parameters. That makes it useful when the objective is non-differentiable, only available as a black box, or expensive to differentiate — a capability optax doesn't currently offer.

## Design

It follows the `obj_fn` / `GradientTransformationExtraArgs` convention already established by `hutchinson_estimator_diag_hessian` (the objective is passed at update time via the `obj_fn` keyword, and the incoming `grads` are ignored):

- **`spsa_gradient(c=0.1, gamma=0.101, seed=None)`** — returns the SPSA gradient estimate. At step *k* it draws a ±1 perturbation `Δ_k`, evaluates `f(θ ± c_k Δ_k)` with `c_k = c / (k+1)**gamma`, and returns `((f₊ − f₋) / (2 c_k)) · Δ_k` (using `1/Δᵢ = Δᵢ` for `Δᵢ ∈ {−1,+1}`).
- **`spsa(learning_rate, ...)`** — convenience optimizer chaining the estimator with a step size (`optax.scale_by_learning_rate`). Spall's classic decaying `a_k` can be supplied via a schedule.

```python
obj_fn = lambda params: loss_fn(params, batch)
opt = optax.contrib.spsa(learning_rate=0.1)
state = opt.init(params)
updates, state = opt.update(grads, state, params, obj_fn=obj_fn)  # grads unused
params = optax.apply_updates(params, updates)
```

## Tests

Dedicated `_spsa_test.py`:
- Convergence on a convex objective (scalar **and** pytree params), using only objective values (grads passed as zeros to prove they're unused).
- **Unbiasedness**: for a quadratic the SPSA estimate is exactly unbiased, so the mean over many draws recovers the true gradient (`atol=5e-2`).
- State/count handling and `ValueError` on missing `obj_fn`/`params`.

All tests pass. It is gradient-free (ignores incoming grads and needs `obj_fn`), so it lives in its own test file rather than the gradient-based `_common_test.py` harness — happy to wire it into `_common_test.py` too if you'd prefer.

### CODE REVIEW

**Comment by jaideeppyne** [comment]
@googlebot I signed it!
